core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -266,7 +282,7 @@ private[deploy] class Worker(
             if (registerMasterFutures != null) {
               registerMasterFutures.foreach(_.cancel(true))
             }
-            val masterAddress = masterRef.address
+            val masterAddress = masterAddressToConnect.get
```

Comment:
How about we conf protect this change (with a default that still uses `masterRef`). If we can merge `master` and `masterAddress` as I suggested above, we can just add a conf on the master and the worker code can be largely unaffected.